packages/nvidia_nat_mcp/src/nat/plugins/mcp/client_impl.py

```diff
@@ -45,6 +43,20 @@ def __init__(self, *args, **kwargs):
         self._mcp_client_server_name: str | None = None
         self._mcp_client_transport: str | None = None
 
+        # Session management attributes for per-session isolation
+        self._session_clients: dict[str, MCPBaseClient] = {}
+        self._session_last_activity: dict[str, datetime] = {}
+        self._session_ref_counts: dict[str, int] = {}  # Track active operations per session
+        self._cleanup_lock = asyncio.Lock()
+        self._session_creation_lock = asyncio.Lock()
```

Comment:
I'm wondering if we could simplify this to use something like a [`aiorwlock.RWLock`](https://github.com/aio-libs/aiorwlock?tab=readme-ov-file) since we have three cases:
- creation of new sessions/resources
- deletion of expired sessions/resources
- access of sessions/resources (happy path)

A meaningful name could be `_session_rwlock`.

The happy path allows for lots of concurrency (multiple readers).

Creation or deletion mandate exclusive (write) access.

Additionally, I believe a per-session `asyncio.Lock` would be required when updating the time/count of each session, since a single session could theoretically be issuing lots of simultaneous requests.

Reply:
ooh...i do like the idea of multi-reader locks.

Reply:
love this suggestion, implemented